File: packages/agent-ui/src/protocol/sourceToolCatalog.ts
import { create } from "@bufbuild/protobuf";

import {
  ToolIdentitySchema,
  type ToolIdentity,
} from "@do-worker/proto/agent_workbench/v2/tool_pb";

interface SourceToolDescriptor {
  semanticKey: string;
  schemaVersion: string;
}

const sourceToolCatalog = new Map<string, ReadonlyMap<string, SourceToolDescriptor>>([
  [
    "acp",
    new Map([
      ["Bash", { semanticKey: "shell.execute", schemaVersion: "1" }],
      ["Read", { semanticKey: "filesystem.read", schemaVersion: "1" }],
      ["Write", { semanticKey: "filesystem.write", schemaVersion: "1" }],
      ["Edit", { semanticKey: "filesystem.edit", schemaVersion: "1" }],
      ["Grep", { semanticKey: "filesystem.search", schemaVersion: "1" }],
      ["WebFetch", { semanticKey: "web.fetch", schemaVersion: "1" }],
      [
        "AskUserQuestion",
        { semanticKey: "interaction.question", schemaVersion: "1" },
      ],
      ["shell", { semanticKey: "shell.execute", schemaVersion: "1" }],
      ["fileChange", { semanticKey: "filesystem.change", schemaVersion: "1" }],
      [
        "image_generation",
        { semanticKey: "media.image.generate", schemaVersion: "1" },
      ],
    ]),
  ],
  [
    "claude",
    new Map([
      ["Bash", { semanticKey: "shell.execute", schemaVersion: "1" }],
      ["Read", { semanticKey: "filesystem.read", schemaVersion: "1" }],
      ["Write", { semanticKey: "filesystem.write", schemaVersion: "1" }],
      ["Edit", { semanticKey: "filesystem.edit", schemaVersion: "1" }],
      ["Grep", { semanticKey: "filesystem.search", schemaVersion: "1" }],
      ["WebFetch", { semanticKey: "web.fetch", schemaVersion: "1" }],
      [
        "AskUserQuestion",
        { semanticKey: "interaction.question", schemaVersion: "1" },
      ],
    ]),
  ],
  [
    "codex",
    new Map([
      ["Bash", { semanticKey: "shell.execute", schemaVersion: "1" }],
      ["Read", { semanticKey: "filesystem.read", schemaVersion: "1" }],
      ["Write", { semanticKey: "filesystem.write", schemaVersion: "1" }],
      ["shell", { semanticKey: "shell.execute", schemaVersion: "1" }],
      ["fileChange", { semanticKey: "filesystem.change", schemaVersion: "1" }],
      [
        "image_generation",
        { semanticKey: "media.image.generate", schemaVersion: "1" },
      ],
    ]),
  ],
  [
    "web-user",
    new Map([
      ["web_search_call", { semanticKey: "web.search", schemaVersion: "1" }],
      [
        "file_search_call",
        { semanticKey: "filesystem.search", schemaVersion: "1" },
      ],
      [
        "code_interpreter_call",
        { semanticKey: "code.interpret", schemaVersion: "1" },
      ],
      ["computer_call", { semanticKey: "computer.use", schemaVersion: "1" }],
      [
        "image_generation_call",
        { semanticKey: "media.image.generate", schemaVersion: "1" },
      ],
      ["mcp_call", { semanticKey: "mcp.call", schemaVersion: "1" }],
      ["mcp_list_tools", { semanticKey: "mcp.list-tools", schemaVersion: "1" }],
    ]),
  ],
]);

export function resolveSourceTool(
  sourceProtocol: string,
  exactSourceToolName: string,
): ToolIdentity | undefined {
  const descriptor = sourceToolCatalog.get(sourceProtocol)?.get(exactSourceToolName);
  if (!descriptor) {
    return undefined;
  }

  return create(ToolIdentitySchema, {
    namespace: `agentsmesh.${sourceProtocol}`,
    semanticKey: descriptor.semanticKey,
    schemaVersion: descriptor.schemaVersion,
    sourceToolName: exactSourceToolName,
  });
}

File: runner/internal/workbench/source_tool_catalog.go
package workbench

import (
	"strings"

	agentworkbenchv2 "github.com/l8ai-cn/agentcloud/proto/gen/go/agent_workbench/v2"
)

type toolDescriptor struct {
	semanticKey string
	category    string
}

var commonTools = map[string]toolDescriptor{
	"Bash":                       {"shell.execute", "shell"},
	"shell":                      {"shell.execute", "shell"},
	"Read":                       {"filesystem.read", "filesystem"},
	"Write":                      {"filesystem.write", "filesystem"},
	"Edit":                       {"filesystem.edit", "filesystem"},
	"fileChange":                 {"filesystem.change", "filesystem"},
	"Grep":                       {"filesystem.search", "filesystem"},
	"WebFetch":                   {"web.fetch", "web"},
	"AskUserQuestion":            {"interaction.question", "interaction"},
	"image_generation":           {"media.image.generate", "media"},
	"web_search_call":            {"web.search", "web"},
	"file_search_call":           {"filesystem.search", "filesystem"},
	"computer_call":              {"computer.use", "computer"},
	"mcp_call":                   {"mcp.call", "mcp"},
	"mcp_list_tools":             {"mcp.list-tools", "mcp"},
	"code_interpreter_call":      {"code.interpret", "code"},
	"image_generation_call":      {"media.image.generate", "media"},
	"workbench.publish_artifact": {"artifact.publish", "artifact"},
}

func resolveToolIdentity(
	sourceProtocol, toolName string,
) (*agentworkbenchv2.ToolIdentity, string, bool) {
	descriptor, ok := commonTools[toolName]
	if !ok {
		return nil, "", false
	}
	namespace := "agentcloud." + sourceProtocol
	if toolName == "workbench.publish_artifact" {
		namespace = "agentcloud.runner"
	}
	return &agentworkbenchv2.ToolIdentity{
		Namespace:      namespace,
		SemanticKey:    descriptor.semanticKey,
		SchemaVersion:  "1",
		SourceToolName: stringPointer(toolName),
	}, descriptor.category, true
}

func sourceProtocol(adapterID string) string {
	switch {
	case strings.HasPrefix(adapterID, "codex"):
		return "codex"
	case strings.HasPrefix(adapterID, "claude"):
		return "claude"
	default:
		return "acp"
	}
}
